res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+//      * Cycle Temporange
+//      * Beat Sync
+//      * Hot Cue Mode
+//      * Beat Loop Mode
+//      * Beat Jump Mode
+//      * Sampler Mode
+//
+//  Custom (Mixxx specific mappings):
+//      * BeatFX: Assigned Effect Unit 1
+//                < LEFT selects EFFECT1
+//                > RIGHT selects EFFECT2
+//                v FX_SELECT selects EFFECT3.
+//                v again to get back to wet/dry mix
+//                ON/OFF toggles selected effect slot
+//                SHIFT + ON/OFF disables all three effect slots.
+//                SHIFT + < selects previous effect
+//                SHIFT + > selects next effect
+//
+//      * 32 beat jump forward & back (Shift + </> CUE/LOOP CALL arrows)
+//      * Toggle quantize (Shift + channel cue)
+//
+//  Not implemented (after discussion and trial attempts):
+//      * Loop Section:
+//        * -4BEAT auto loop (hacky---prefer a clean way to set a 4 beat loop
+//                            from a previous position on long press)
+//
+//        * CUE/LOOP CALL - memory & delete (complex and not useful. Hot cues are sufficient)
+//
+//      * Secondary pad modes (trial attempts complex and too experimental)
+//        * Keyboard mode
+//        * Pad FX1
+//        * Pad FX2
+//        * Keyshift mode
+
+var PioneerDDJ400 = {};
+
+PioneerDDJ400.lights = {
+    beatFx: {
+        status: 0x94,
+        data1: 0x47,
+    },
+    deck1: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x90,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x90,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x90,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x90,
+            data1: 0x48,
+        },
+    },
+    deck2: {
+        vuMeter: {
+            status: 0xB0,
+            data1: 0x02,
+        },
+        playPause: {
+            status: 0x91,
+            data1: 0x0B,
+        },
+        shiftPlayPause: {
+            status: 0x91,
+            data1: 0x47,
+        },
+        cue: {
+            status: 0x91,
+            data1: 0x0C,
+        },
+        shiftCue: {
+            status: 0x91,
+            data1: 0x48,
+        },
+    },
+};
+
+// Store timer IDs
+PioneerDDJ400.timers = {};
+
+// Stores padmode state and encapsulates LED logic
+PioneerDDJ400.performancePads = {
+    modes: {
+        HOTCUE: 0x1B,
+        BEATLOOP: 0x6D,
+        BEATJUMP: 0x20,
+        SAMPLER: 0x22,
+        KEYBOARD: 0x69,
+        PADFX1: 0x1E,
+        PADFX2: 0x6B,
+        KEYSHIFT: 0x6F
+    },
+
+    // Start in hotcue mode
+    state: [
+        0x1B, // Channel 1
+        0x1B  // Channel 2
+    ],
+
+    // Toggles the shift specific lights for the current mode
+    toggleShiftLights: function(channel, value) {
+        if (this.state[channel] === this.modes.BEATJUMP) {
```

Comment:
Is this only needed for the beatjump pad mode? Or have you not yet implemented it for the others?

Reply:
Only needed for beatjump mode. I'll see if this can be simplified. Not entirely happy with it either

Reply:
So this could in theory be used to implement Keyboard and Key Shift modes at a later stage (i.e., probably never :wink: ).

Out of all the pad modes, they're the only ones that (like beatjump) light up in feature specific ways when shift is pressed.

Reply:
I'm unclear what LEDs this toggles. Why is it only needed for beatjump mode but not the other implemented modes?

Reply:
Was very happy to clean all of this out and just implement the LEDs as outputs. The script is now very straightforward and uniform in its implementation.